### PR TITLE
OCPBUGSM-37106: List only physical disks on inventory.

### PR DESCRIPTION
**Pull request by eliorerz**
List only physical disks on inventory. Ignore LVM and other volumes.

### CODE REVIEW

File: src/inventory/disks.go [hunk 1]
```diff
@@ -179,6 +179,29 @@ func checkEligibility(disk *ghw.Disk) (notEligibleReasons []string, isInstallati
 	return notEligibleReasons, isInstallationMedia
 }
 
+func (d *disks) getPhysicalDisks() map[string]string {
+	stdout, stderr, exitCode := d.dependencies.Execute("lsblk", "-o", "TYPE,KNAME")
```
Comment (osherdp) [inline src/inventory/disks.go:183]: Why won't we use information gathered by ghw here instead?
It already chroot to the location /host/ which has the more relevant data about the host (instead of risking getting data from the container instead)
Also less parsing is needed

File: src/inventory/disks.go [hunk 2]
```diff
@@ -179,6 +179,29 @@ func checkEligibility(disk *ghw.Disk) (notEligibleReasons []string, isInstallati
...
+	if exitCode != 0 {
+		logrus.Warnf("Could not get physical disks information: %s", stderr)
+		return nil
+	}
+
+	lines := strings.Split(stdout, "\n")
+	physicalDisks := map[string]string{}
+	for lineno, line := range lines {
+		if lineno > 0 {
```
Comment (tsorya) [inline src/inventory/disks.go:192]: you can use "lsblk -n -o TYPE,KNAME" and then there will be no need for  "if lineno > 0"

File: src/inventory/disks.go [hunk 3]
```diff
@@ -194,11 +217,16 @@ func (d *disks) getDisks() []*models.Disk {
 	}
 
 	diskPath2diskWWN := d.getDisksWWNs()
+	physicalDisks := d.getPhysicalDisks()
```
Comment (osherdp) [inline src/inventory/disks.go:220]: Just my phone or the indentation here looks weird?

File: src/inventory/disks.go [hunk 4]
```diff
@@ -179,6 +179,29 @@ func checkEligibility(disk *ghw.Disk) (notEligibleReasons []string, isInstallati
...
+	}
+
+	lines := strings.Split(stdout, "\n")
+	physicalDisks := map[string]string{}
+	for lineno, line := range lines {
+		if lineno > 0 {
+			data := strings.Split(line, " ")
+			volumeType := data[0]
+			volumeName := data[len(data)-1]
+			if volumeType == "disk" {
```
Comment (tsorya) [inline src/inventory/disks.go:196]: i am not sure how removable disks will appear here, maybe we should check lvm and ignore by it?

File: src/inventory/disks.go [hunk 5]
```diff
@@ -179,6 +179,29 @@ func checkEligibility(disk *ghw.Disk) (notEligibleReasons []string, isInstallati
 	return notEligibleReasons, isInstallationMedia
 }
 
+func (d *disks) getPhysicalDisks() map[string]string {
+	stdout, stderr, exitCode := d.dependencies.Execute("lsblk", "-o", "TYPE,KNAME")
+	if exitCode != 0 {
+		logrus.Warnf("Could not get physical disks information: %s", stderr)
+		return nil
```
Comment (osherdp) [inline src/inventory/disks.go:186]: nil allows for the "is key in map" check below?
Worth also adding a UT for this case

Comment (osherdp) [reply src/inventory/disks.go:186]: Or rather not allow failures

**Review by osherdp — CHANGES_REQUESTED** [review]
Just those couple of comments, and add UTs for some cases

File: src/inventory/disks.go (hunk 3 above)
Comment (omertuc) [reply src/inventory/disks.go:220]: Just your phone. GitHub is only usable on mobile when the browser is in Desktop mode

File: src/inventory/disks.go (hunk 1 above)
Comment (omertuc) [reply src/inventory/disks.go:183]: I agree with Osher, if it's possible to use ghw here we better do that.

But if we're doing lsblk, let's use `lsblk --json --output TYPE,KNAME` for more structured parsing

File: src/inventory/disks.go [hunk 6]
```diff
@@ -179,6 +179,29 @@ func checkEligibility(disk *ghw.Disk) (notEligibleReasons []string, isInstallati
...
+
+	lines := strings.Split(stdout, "\n")
+	physicalDisks := map[string]string{}
+	for lineno, line := range lines {
+		if lineno > 0 {
+			data := strings.Split(line, " ")
+			volumeType := data[0]
+			volumeName := data[len(data)-1]
+			if volumeType == "disk" {
+				physicalDisks[volumeName] = volumeType
```
Comment (omertuc) [inline src/inventory/disks.go:197]: physicalDisks is being used as a Set structure so please just give it a "true/false" boolean value (which would always be true) rather than a "volumeType" value

File: src/inventory/disks.go (hunk 4 above)
Comment (omertuc) [reply src/inventory/disks.go:196]: >maybe we should check lvm and ignore by it?
This will not suffice. I device-mapper devices appearing with the "crypt" type by lsblk

Comment (omertuc) [reply src/inventory/disks.go:196]: Looking at lsblk source code maybe even checking for "disk" won't be good enough, for example it looks like RAID arrays (which should technically be fine for us to install on) may appear as `raid` instead of `disk`:

https://github.com/util-linux/util-linux/blob/4939964c9d1977522de1a3de03ebb3011ca866de/lib/blkdev.c#L355-L356

Comment (omertuc) [reply src/inventory/disks.go:196]: From my local testing with ramdisks, swap, LVM, encrypted partitions I find this method quite reliable:

```
if strings.HasPrefix(disk.Name, "dm-") ||
strings.HasPrefix(disk.Name, "loop") ||
strings.HasPrefix(disk.Name, "md") {
	// Ignore this disk!
}
```

It's pretty much identical to [what](https://github.com/util-linux/util-linux/blob/4939964c9d1977522de1a3de03ebb3011ca866de/misc-utils/lsblk.c#L413-L455) `lsblk` is doing behind the scenes when determining the `type`, except without the extra problematic transformations it does with `[blkdev_scsi_type_to_name](https://github.com/util-linux/util-linux/blob/4939964c9d1977522de1a3de03ebb3011ca866de/lib/blkdev.c#L332)` in the end  

I noticed that it doesn't detect the default swap blocked device used by Fedora with [SwapOnZRAM](https://fedoraproject.org/wiki/Changes/SwapOnZRAM)

Even `lsblk` shows that device as "disk" even though it's definitely a virtual block device, and the `HasPrefix` comparisons above won't detect it either. The issue is that such block devices can be created with any name - you can try it yourself with `mknod -m 660 /dev/blabla b 1 0`

One particular `lsblk` column that stands out in being especially good at detecting "real" disks is the `subsystems` column - it seems that any block device is not backed up by real physical connections will have the value `block` in `subsystems`

Comment (omertuc) [reply src/inventory/disks.go:196]: It seems like `ghw`'s `BusPath` being equal to `unknown` is a pretty good indication that it's not a true physical disk, this `BusPath` `ghw` field, taken from udev's `ID_PATH` is pretty similar to the `subsystems` column calculated by `lsblk`. I think we should just look at it and filter any disks that don't have it.

File: src/inventory/disks.go (hunk 1 above)
Comment (omertuc) [reply src/inventory/disks.go:183]: >Although I don't think ghw has the information we're looking for

Or maybe it does, see [this](https://github.com/openshift/assisted-installer-agent/pull/259/files#r760629845) comment

Comment (eliorerz) [reply src/inventory/disks.go:183]: ghw doesn't give any information about LVK volumes

File: src/inventory/disks.go (hunk 4 above)
Comment (eliorerz) [reply src/inventory/disks.go:196]: There are other types we want to ignore. Currently we supports only physical volumes so I thought it will be better not to ignore every unwanted type.
Also, removable disk type is `disk` and will not be filtered here

File: src/inventory/disks.go (hunk 3 above)
Comment (eliorerz) [reply src/inventory/disks.go:220]: I think it's just your phone :)

File: src/inventory/disks.go (hunk 4 above)
Comment (eliorerz) [reply src/inventory/disks.go]: > It appears ghw's BusPath being equal to unknown is a pretty good indication that it's not a true physical disk

Ghw unknown fields are not fully indicates that this volume is physical or not (at least not explicitly).

> From my local testing with ramdisks, swap, LVM and encrypted partitions I find this method quite reliable

Same here, the prefix of the device name is not good enough for deciding if it physical of not

Comment (omertuc) [reply src/inventory/disks.go]: >Ghw unknown fields are not fully indicates that this volume is physical or not (at least not explicitly).

Do you have an example where a physical disk has a ghw BusPath value that is `"unknown"`?

Comment (omertuc) [reply src/inventory/disks.go]: >Same here, the prefix of the device name is not good enough for deciding if it physical of not

It's literally what `lsblk` is doing as I've [linked](https://github.com/util-linux/util-linux/blob/4939964c9d1977522de1a3de03ebb3011ca866de/misc-utils/lsblk.c#L413-L455) in the message above. Except `lsblk` might call RAID disks "raid" rather than "disk" which will cause you to wrongfully filter them out, as mentioned

File: src/inventory/disks.go (hunk 1 above)
Comment (omertuc) [reply src/inventory/disks.go]: If the `BusPath` is good indication that a device is virtual (`"unknown"` bus types indicates it's not backed up by physical connections) then `ghw` might not give us information about LVK volumes specifically - but it gives us something even more general and useful which is ignoring all non physical disks

Comment (omertuc) [reply src/inventory/disks.go]: Also how do you overcome the `/host` problem @osherdp mentioned?

File: src/inventory/disks.go (hunk 4 above)
Comment (eliorerz) [reply src/inventory/disks.go]: > Do you have an example where a physical disk has a ghw BusPath value that is "unknown"?

No, But do you have a list of all the volumes and ghw BusPath is not "unknown"? :) 
Although you might be right, I think for proving that you need to check all the options for all physical disks and configurations. It's much easier to get the physical disks list then assume that ghw BusPath set to "unknown" is what we wanted (unless it was specified that "unknown"  specify unphysical disk) .

>Same here, the prefix of the device name is not good enough for deciding if it physical of not
It's literally what lsblk is doing as I've linked in the message above. Except lsblk might call RAID disks "raid" rather than "disk" which will cause you to wrongfully filter them out, as mentioned

That a valid case.  We might want to specify the specific physical disk types that we currently supporting instead of filtering by BusPath.

Comment (omertuc) [reply src/inventory/disks.go]: > No, But do you have a list of all the volumes and ghw BusPath is not "unknown"? :)
No I do not. I tried looking on my local machine and some old inventories I had saved and it seemed pretty reliable.

It also makes sense - the bus path tells you how the disk is physically connected, a disk that exists only virtually isn't physically connected in any way. 

Choosing to look at `lsblk` `type` output is also "guesswork" where we have no guarantees it will work 100% of the time, except I've provided examples (SwapOnZRAM) where the `lsblk` `type` method falls short and causes you to report fake disks. 

> That a valid case. We might want to specify the specific physical disk types that we currently supporting instead of filtering by BusPath.

Or just use the prefix comparisons that I provided which are identical to what lsblk is doing before it starts to potentially split the 'disk' type into other more specific types. If what `lsblk` is doing to calculate the `type` is so simple, why even use `lsblk`? It complicates things, requires parsing, requires us to compare to our already-existing list of disks which we got with a `chroot` that is necessary.

While we could instead just do 3 HasPrefix calls.

In summary there are 3 solutions here:
1) lsblk
2) hasprefix
3) buspath

I don't see how (1) along with:

> We might want to specify the specific physical disk types that we currently supporting 

is in any way better than (2).

and regardless of (2) being simpler and more robust than (1), I think we should just go with (3) because I couldn't find an example where it's wrong while I could find an example where (1) and (2) are wrong

Comment (eliorerz) [reply src/inventory/disks.go]: How's `hasPrefix` is good? Are we familiar with all the prefixes? You need to maintain list of prefixes? Are you sure we know them all? You cannot conclude about the all from few test cases.

Regarding 3 (bustype), I agree that this a good approach, but I'm still feeling that this might be more open to bugs than lsblk. We can combine 1 and 3 together to remove some unwanted disks (such as zram)

Comment (eliorerz) [reply src/inventory/disks.go]: If lsblk is a big no no from your perspective I believe we can use bustype as an indication for physical (although I'm not completely comfortable with it)

Comment (omertuc) [reply src/inventory/disks.go]: > How's hasPrefix is good? 

Because it's exactly what `lsblk` is doing anyway. `lsblk` is simply comparing those 3 prefixes before saying the type is `disk`, then it goes on to split `disk` into even more types which is useless and complicated things further from our perspective. You can go look at the source code, I've linked it multiple times already.

> Are we familiar with all the prefixes? You need to maintain list of prefixes? Are you sure we know them all? 

When you "use lsblk" you're just delegating those responsibilities to `lsblk`. Are the `lsblk` maintainers familiar with all the prefixes? Do they maintain a list of prefixes? Maybe, maybe not. Maybe it's a good idea delegating that responsibility to `lsblk`. But that delegation has a cost - it's putting too much complexity between us and what we want to achieve (know if disk is virtual or not) into a program with no guaranteed stable interface:
How do we know they won't completely change how they determine the `type`, how do we know they don't introduce even more granular types a year from now? Who decided that `type == "disk"` is even the right thing to do in order to determine if a disk is virtual or not? I've even shown it's **not** the right thing in case on SwapOnZRAM 

Comparing these 3 prefixes is simply saying "Let's ignore device-mapper block devices, loop devices, and multiple-devices block devices", it's much more clear and stable then delegating to `lsblk` which have no guaranteed stability, already returns types that are too granular (raid vs disk), and might even become more granular in the future outside of our control, and is a black box until you open up the source code and try to follow what it's actually doing to determine the `type`.

> You cannot conclude about the all from few test cases.

You also cannot conclude that looking at `lsblk` `type` is the right thing to do from a few test cases 

> that this might be more open to bugs than lsblk

I don't understand how BusPath is more open to bugs than `lsblk` `type` in any way.

> If lsblk is a big no no from your perspective I believe we can use bustype as an indication for physical (although I'm not completely comfortable with it)

`lsblk` is a big no until you explain / address the [chroot problem](https://github.com/openshift/assisted-installer-agent/pull/259#discussion_r760970629). Maybe it's an issue, maybe it's not, but it needs to be addressed and explained why it's not an issue.

Once that's addressed it's possible to consider it, but given all I've said about it, I don't understand what benefits it has that the `HasPrefix` method doesn't have and I don't understand what downsides other methods have that the `lsblk` `type` method doesn't have.

Comment (eliorerz) [reply src/inventory/disks.go]: >lsblk is simply comparing those 3 prefixes before saying the type is disk 
....
>When you "use lsblk" you're just delegating those responsibilities to lsblk. Are the lsblk maintainers familiar with all the prefixes? Do they maintain a list of prefixes? Maybe, maybe not. Maybe it's a good idea delegating that responsibility to lsblk.


As you mentioned, lsblk is comparing the prefixes to decide whether it's a physical disk or not so I believe they are familiar with those prefixes. Also, I believe that lsblk are more common used that assisted-service (for now) and more likely that for every change in the prefixes the fix will apply before we even noticed that.

File: src/inventory/disks.go (hunk 1 above)
Comment (eliorerz) [reply src/inventory/disks.go]: lsblk is working the same inside the container and on the host itself

File: src/inventory/disks.go (hunk 4 above)
Comment (omertuc) [reply src/inventory/disks.go]: >  Also, I believe that lsblk are more common used that assisted-service (for now) and more likely that for every change in the prefixes the fix will apply before we even noticed that.

I would completely agree with you here and with that approach *if* `lsblk` had a boolean "is_real_disk" column.

But the `type` column doesn't have that meaning, it's just some column that happens to *currently* fit the distinction you're looking for under *some* circumstances when its value equals `disk`. This coincidental behavior is not stable and I don't think relying on it is a good idea. The `lsblk` maintainers give no guarantees here and are completely free to change this behavior in ways that can easily break our flow

File: src/inventory/disks.go (hunk 1 above)
Comment (omertuc) [reply src/inventory/disks.go]: I assume it works because of this mount https://github.com/openshift/assisted-service/blob/e125b59fd271dbeace653e02631a03d8d57ba98f/internal/host/hostcommands/inventory_cmd.go#L37 and some other default mounts 

But we've recently made this `chroot` configurable due to some new projects using the agent code, using `lsblk` without going through ghw would mean that this configuration is ignored, which is not ideal (although I don't think it'll break any current real world use cases)

File: src/inventory/disks.go (hunk 4 above)
Comment (eliorerz) [reply src/inventory/disks.go]: OK, So in that case, what is your suggestion? Use `BusPath` or `HasPrefix` instead ?

Comment (omertuc) [reply src/inventory/disks.go]: I say let's do `HasPrefix` for now because it's very unlikely to get rid of disks that are actually real.

But we should open a ticket to look into the reliability of `BusPath`, because if we can't find counter-examples for where it doesn't work, it's a really simple and nice solution that makes sense and doesn't require maintaining a disallow-list of "fake disk technologies" (loop, dm, md).

Comment (eliorerz) [reply src/inventory/disks.go]: OK, I'll use `HasPrefix` to filter the unphysical disks.
Note that I'm also adding zram to the filtered disks so it will not be listed in the final disks list.
And on other subject, I noticed that ghw is not always identifying if device is removable or nor (I reproduced both cases in my personal laptop), we might need to open ticket for that as well.

File: src/inventory/disks.go [hunk 7]
```diff
@@ -179,6 +179,13 @@ func checkEligibility(disk *ghw.Disk) (notEligibleReasons []string, isInstallati
 	return notEligibleReasons, isInstallationMedia
 }
 
+func IsPhysicalDisk(disk *block.Disk) bool {
+	return !(strings.HasPrefix(disk.Name, "dm-") ||
```
Comment (omertuc) [inline src/inventory/disks.go:183]: Let's add a small description about each technology? Just to make it easier to google what each one is about

```go
// Kernel device-mapper block devices, technology used LVM/Docker and others to create virtual disks
// Loop devices (see `man loop`)
// Default name usually assigned to "swap on ZRAM" block devices
// Linux multiple-device-driver block devices
```
